documentation/specs/static-graph.md

```diff
@@ -143,49 +139,100 @@ This is a proposal for what the public API may look like:
 
-After execution, the results of the export targets will be cached and the project can be disposed. Note that this gives project execution a concrete start and end, while previously projects needed to remain active just in case any other project needed to execute a target on it.
+A project reference protocol could contain multiple targets, for example $A \to B, A$. This means that building $A$ on the parent will lead to $B$ and $A$ getting called on the children. If all nodes in the graph repeat the same rule, then the rule is repeated recursively on all nodes (this would be a regular operation). However, a node can choose to implement the protocol differently, which would lead to an irregular protocol. In the following example, the entry targets are:
+- Proj4 is built with targets $B, A, C, D$. On multiple parents, the incoming targets get concatenated. The order does not matter, as MSBuild has non-deterministic p2p ordering.
+- Proj3 and Proj2 build with $B, A$, as specified by the rule in Proj1.
+- Proj1 builds with A, because it's the root of the graph.
 
-For projects which depend on other projects, the MSBuild task will look up the target result in the cache. If the target is missing from the cache, an error will be logged and the build will fail. If the project is calling into itself either via `CallTarget` or via `MSBuild` with a different set of global properties, this will be always allowed (support for cross-targeting).
+```
+            A+-->Proj1   A->B, A
+                 /    \
+           B, A /      \ B, A
+               /        \
+              v          v
+  A->B, A   Proj3      Proj2   A->C, D
+             \            /
+        B, A  \          /  C, D
+               \        /
+                \      /
+                 v    v
+                 Proj4   A->B, A
+```
+
+The common project reference protocols (Build, Rebuild, Restore, Clean) will be specified by the common props and targets file in the msbuild repository. Other SDKs can implement their own protocols (e.g. ASPNET implementing Publish).
+
+Here are the rules for the common protocols:
+
+$Build \to GetTargetFrameworks, \langle default \rangle, GetNativeManifest, GetCopyToOutputDirectoryItems$
+
+The default target (represented here as $\langle default \rangle$) is resolved for each project.
+
+$Clean \to GetTargetFrameworks, Clean$
+
+$Rebuild \to \$Clean, \$Build$
+
+$Rebuild$ maps to two referenced protocols. $\$Clean$ and $\$Build$ get substituted with the contents of the $Clean$ and $Build$ protocols.
+
+Restore is a composition of two rules:
+- $Restore \to \_IsProjectRestoreSupported, \_GenerateRestoreProjectPathWalk, \_GenerateRestoreGraphProjectEntry$
```

Comment:
Is there a benefit in representing Restore in this way? Can restore not just continue to use the JIT reference approach, or potentially be rewritten to request the graph and do its own walk based on that?

Reply:
Yes, once we have the graph APIs it can be rewritten. But until then, if we want to run restore on the static graph, we'll need to be able to represent it's protocol in order to infer entry targets.